examples/singlestep-event-example.c

```diff
@@ -82,23 +87,91 @@ int main (int argc, char **argv)
+    atexit(&exit_cleanup);
+
+    // get number of vcpus
+    unsigned int num_vcpus = vmi_get_num_vcpus(vmi);
 
-    //Single step setup
+    // Single step setup
     memset(&single_event, 0, sizeof(vmi_event_t));
     single_event.version = VMI_EVENTS_VERSION;
     single_event.type = VMI_EVENT_SINGLESTEP;
     single_event.callback = single_step_callback;
     single_event.ss_event.enable = 1;
     SET_VCPU_SINGLESTEP(single_event.ss_event, 0);
-    vmi_register_event(vmi, &single_event);
-    while (!interrupted ) {
+
+    // register
+    if (VMI_FAILURE == vmi_register_event(vmi, &single_event)) {
+        fprintf(stderr, "Failed to register singlestep event\n");
+        return 1;
+    }
+
+    // event loop
+    while (!interrupted) {
         printf("Waiting for events...\n");
-        vmi_events_listen(vmi,500);
+        if (VMI_FAILURE == vmi_events_listen(vmi,500)) {
+            fprintf(stderr, "Failed to listen on events\n");
+            return 1;
+        }
     }
-    printf("Finished with test.\n");
 
-    // cleanup any memory associated with the libvmi instance
-    vmi_destroy(vmi);
+    // toggling singlestep off/on
+
+    // pause before cleaning the ring
+    // this prevents new events from being queued
+    if (VMI_FAILURE == vmi_pause_vm(vmi)) {
+        fprintf(stderr, "Failed to pause VM\n");
+        return 1;
+    }
+    // clean event ring
+    if (VMI_FAILURE == vmi_events_listen(vmi, 0)) {
+        fprintf(stderr, "Failed to pause VM\n");
+        return 1;
+    }
+
+    // disable singlestep on all vcpus
+    for (unsigned int vcpu=0; vcpu < num_vcpus; vcpu++) {
+        if (VMI_FAILURE == vmi_toggle_single_step_vcpu(vmi, &single_event, vcpu, false)) {
+            fprintf(stderr, "Failed to stop singlestepping on VCPU %d\n", vcpu);
+            return 1;
+        }
+    }
+    printf("Singlestep stopped\n");
+    if (VMI_FAILURE == vmi_resume_vm(vmi)) {
+        fprintf(stderr, "Failed to resume VM\n");
+        return 1;
+    }
+
+    // VM should be running
+    // sleep(5);
+
+    // toggle singlestep back ON
+    // pause before changing VM state
+    if (VMI_FAILURE == vmi_pause_vm(vmi)) {
+        fprintf(stderr, "Failed to pause VM\n");
+        return 1;
+    }
+    printf("Restarting singlestep\n");
+    for (unsigned int vcpu=0; vcpu < num_vcpus; vcpu++) {
+        if (VMI_FAILURE == vmi_toggle_single_step_vcpu(vmi, &single_event, vcpu, true)) {
+            fprintf(stderr, "Failed to enable singlestep on VCPU %d\n", vcpu);
+            return 1;
+        }
+    }
+    if (VMI_FAILURE == vmi_resume_vm(vmi)) {
+        fprintf(stderr, "Failed to resume VM\n");
+        return 1;
+    }
+    // process a few more singlestep events
+    for (int i=0; i < 5; i++) {
+        printf("Waiting for events...\n");
+        if (VMI_FAILURE == vmi_events_listen(vmi,500)) {
+            fprintf(stderr, "Failed to listen on events\n");
+            return 1;
+        }
+    }
+    printf("Finished with test.\n");
```

Comment:
Perhaps add a comment that all singlestep events are disabled by `vmi_destroy`.

Reply:
Done